### DISCUSSION TITLE
Errore Manual Authentication

### DISCUSSION THREAD

**Discussion by damiandominella** [fattureincloud/api#160] [Q&A] [answered]
Ciao a tutti, 
stiamo migrando le api dalla v1 alla v2, la nostra integrazione con fatture in cloud è molto semplice, semplicemente effettuiamo chiamate API da un nostro servizio backend a fatture in cloud.

Stiamo avendo dei problemi con l'autenticazione, utilizziamo il [flusso manuale](https://developers.fattureincloud.it/docs/basics/authentication#-manual-authentication).

Ho correttamente seguito la procedura, creato un app, generato un token valido e riesco correttamente ad utilizzare gli endpoint non company scoped (Get User Info, Get User Companies). 

Dalla lista di companies quindi, ottengo il nostro company id che è necessario per poter utilizzare gli altri endpoint.

Quando però provo ad effettuare delle chiamate API utilizzando il company id, ricevo sempre lo stesso errore:

```
{"error":"unauthorized","error_description":"The given user is not authorized to use this application."}
```

Come se il token si riferisse ad un utente che non ha accessi all'applicazione, anche se la procedura di creazione e collegamento l'ho effettuata correttamente come da guida.

Che devo fare? 

Grazie

**Comment by valmoz** [comment, marked as answer]
Buonasera,
se riceve questo errore le cause possono essere tre:

1. Il company id che sta passando non è corretto (potrebbe aver inserito il parametro errato ad esempio)
2. Il token non è stato abilitato sulla company che state cercando di utilizzare (vi chiederei di controllare la sezione applicazioni collegate -> aziende collegate)
3. state utilizzando un'app privata, ma la company non è in whitelist (anche se mi sembra improbabile)

Per poter indagare più approfonditamente vi suggeriamo di contattare il nostro supporto clienti.

Grazie
Mauro

**Comment by damiandominella** [comment]
Salve, ho verificato i primi due punti citati e:

1. Il company id è corretto (lo ottengo correttamente utilizzando `Get User Companies`.)
2. Il token è abilitato alla company, se accedo alla sezione che mi ha indicato trovo giustamente l'azienda collegata all'applicazione

Riguardo al punto 3, l'app è si privata. Ma come verifico che la company sia in white list ? 

Grazie

**Reply by valmoz** [reply]
Buongiorno @damiandominella 

Per fugare ogni dubbio le chiederei prima di tutto di recuperare il suo token e di provare ad effettuare la chiamata company-scoped tramite la nostra [API Reference](https://developers.fattureincloud.it/api-reference) oppure tramite [Postman](https://developers.fattureincloud.it/docs/quickstarts/postman-collection), in modo tale da verificare se il problema sta nel token o nel codice. [Qui](https://developers.fattureincloud.it/docs/api-reference) spieghiamo come usare l'API Reference.

Per evitare che si verifichino altri problemi, vi suggerirei di provare prima di tutto la chiamata [Get Company Info](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/company/info), che è la più semplice chiamata company-scoped e non richiede parametri aggiuntivi.

vi chiederei comunque di verificare se il company id che avete recuperato con Get User Companies è stato inserito correttamente nella richiesta, solo per evitare questa ipotesi. Capita spesso che per errori nel codice venga inserito per errore un valore null nella richiesta.
Se state usando le nostre SDK il parametro dovrebbe essere impostato automaticamente, mentre se state usando un client HTTP l'id deve essere inserito correttamente come path parameter. Ad esempio:
`/c/12345/company/info`

Il punto tre mi sembra altamente improbabile, in quanto siete già riusciti ad ottenere il company id, e non dovrebbe essere possibile recuperare un company id se non se ne ha il controllo. Le chiedo comunque di verificare, accedendo alla pagina dell'applicazione privata (sezione Sviluppatore) e controllando l'elenco "E-mail degli utenti che possono utilizzare l'applicazione".



Nel caso in cui questi controlli dovessero andare a buon fine, le chiederei di contattare il supporto clienti incollando il link a questa discussione nella richiesta, in modo tale da poter recuperare il suo company id (che non deve essere pubblicato sulla community) e rendere quindi più semplice le indagini sul problema.

Le modalità di contatto sono disponibili qui: [Support & Feature Request](https://developers.fattureincloud.it/docs/support)

Per completezza le reinvio la documentazione alle chiamate [Company scoped](https://developers.fattureincloud.it/docs/basics/company-scoped-methods).

P.S.: Una volta ottenuto, il company-id non cambia per una singola azienda. Quindi sarà necessario recuperarne l'ID solo la prima volta.

Grazie
Mauro

**Reply by damiandominella** [reply]
Stiamo effettuando i test ancor prima di scrivere codice, utilizzando il vostro API reference. 
Ho trovato [questo altro post](https://github.com/fattureincloud/api/discussions/130) , dove però la soluzione per l'utente è stata rendere pubblica la loro applicazione, cosa che non credo sia necessario fare.

Tutti i controlli proposti sono già andati a buon fine.

Ho già contattato il supporto clienti rimarrò in attesa di una risposta.
Chiudo questo topic,

Grazie

**Reply by valmoz** [reply]
Buongiorno @damiandominella,
ho notato che la vostra whitelist era totalmente vuota, quindi era stato rimosso anche il vostro account. Essendo un'app privata, non era più in grado di utilizzare il token generato in quanto relativo ad un account non autorizzato.

Abbiamo provveduto a reinserire il vostro account all'interno della whitelist. Ci faccia sapere se ora funziona tutto correttamente.

Vi ricordo che per l'utilizzo da altri account sarà necessario aggiungere l'account alla whitelist oppure richiedere la modifica della visibilità dell'applicazione.

Grazie
Mauro

**Reply by andrealune** [reply]
@valmoz ho avuto lo stesso problema per i primi 90 minuti dopo aver aperto l'account, poi magicamente ha iniziato a funzionare e dopo qualche minuto non funzionava ancora.
La documentazione è poco chiara, il sistema di autenticazione inutilmente complesso e le Api inconsistenti.

**Es.**
GET https://api-v2.fattureincloud.it/user/companies => Funziona su Api Reference/Postman
```
{
  "data": {
    "id": 1081372
    ...
}
```
GET https://api-v2.fattureincloud.it/c/1081372/info => Funziona solo su Api reference, su Postman restituisce errore indefinito
```
{
    "error": {
        "message": "",
        "code": 0
    }
}
```
**EDIT: Errore mio su POSTMAN, l'endpoint corretto è https://api-v2.fattureincloud.it/c/1081372/company/info**

GET https://api-v2.fattureincloud.it/c/1081372/info/payment_methods => Funziona su Api Reference/Postman
```
{
  "data": [
    {
      "id": 2048886,
      "name": "IBAN",
      "is_default": true,
      "default_payment_account": null
    }
  ]
}
```
ma provando a generare una fattura => POST https://api-v2.fattureincloud.it/c/1081372/issued_documents
```
{"data":{"type":"invoice","use_gross_prices":true,"visible_subject":"Canone annuale d'uso del servizio","entity":{"name":"TEST","address_street":"TEST 11A","address_postal_code":"20133","address_city":"MILANO (MI)","address_province":"MI","country":"Italia"},"currency":{"id":"EUR","symbol":"€"},"language":{"code":"it","name":"Italiano"},"items_list":[{"name":"Canone annuale d'uso del servizio","gross_price":120,"qty":1}],"payments_list":[{"amount":120,"due_date":"2023-01-20","paid_date":"2023-01-20","status":"paid","payment_account":{"id":"2048886"}}]}}
```

restituisce:

```
error: {
    "message": "Conto di pagamento non trovato.",
    "validation_result": null
}
```

**Reply by valmoz** [reply]
Buongiorno @andrealune 
mi spiace che stia avendo problemi con le nostre API, siamo qui per cercare di darvi il supporto necessario per risolvere questi problemi.

Cercherò di risponderle per punti, suggerendole però di non mischiare troppo gli argomenti. Aprendo una nuova discussione, permetterà ad altri utenti in difficoltà di cercare più facilmente una soluzione.

Prima di tutto, partiamo dalla fase di autenticazione, che tra l'altro è la parte più complessa dell'utilizzo delle API.
Noi offriamo principalmente due metodi di autenticazione:
- OAuth 2.0 Authorization Code
- Manual Token
(ne esiste un terzo, ma per semplicità lo ometterò perché è dedicato a casistiche particolari).

Tra i due, il metodo OAuth è sicuramente il metodo più complesso. Le faccio però notare che non ce lo siamo inventato noi, ma è uno [standard](https://oauth.net/2/) utilizzato tra gli altri da Google e che definisce un flusso di autorizzazione sicuro. Ci siamo semplicemente limitati ad adottare lo standard.
Tra l'altro, per semplificare l'utilizzo di OAuth abbiamo preparato delle [guide dedicate](https://developers.fattureincloud.it/docs/authentication/code-flow/) e abbiamo anche messo a disposizione degli Helper all'interno delle nostre SDK.

La buona notizia è che, se non le piace OAuth, mettiamo a disposizione anche il [Token Manuale](https://developers.fattureincloud.it/docs/authentication/manual-authentication), che può essere considerato in svariati modi fuorché difficile.
In pratica per ottenere il token manuale basta:

- Creare un'[applicazione](https://developers.fattureincloud.it/docs/basics/create-an-app) e selezionare il token manuale
- Copiare il Client ID mostrato sulla pagina
- Accedere alla sezione Impostazioni > Applicazioni collegate
- Cliccare su "Collega nuova applicazione"
- Inserire il Client ID della vostra applicazione da collegare
- Seguire la procedura guidata per selezionare la company e i permessi
- Copiare il token da utilizzare

Una volta copiato il token, basta inserirlo nella richiesta come Header e il gioco è fatto. 
Nel caso in cui ci dovessero essere ancora difficoltà, vi chiederei di farci sapere in cosa consistono per poter migliorare il flusso di autenticazione.

Vi risponderò nei commenti successivi per i diversi errori segnalati, per evitare di darvi una risposta troppo lunga.

A disposizione
Mauro

**Reply by andrealune** [reply]
Grazie @valmoz, con sistema di autenticazione complessa mi riferivo al sistema Manual Token che avevo comunque configurato correttamente ed ha iniziato a funzionare solo dopo un determinato tempo.

**Reply by valmoz** [reply]
@andrealune a me non sembra essere una procedura complessa, ma rispetto la sua opinione.

Tornando agli errori sui metodi:
1. vedo il suo edit sul metodo errato, segnalerò che l'errore in caso di metodo errato è poco descrittivo
2. per l'errore "conto di pagamento non trovato", le chiederei di verificare che l'ID del payment_account sia effettivamente esistente e tra quelli appartenenti alla sua company. Può farlo tramite il metodo [List Payment Accounts](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/info/payment_accounts) o [creandone uno nuovo](https://developers.fattureincloud.it/api-reference#post-/c/-company_id-/settings/payment_accounts) a seconda delle sue necessità.

Per semplificare la creazione di una fattura, che è sicuramente complicata e richiede molti step, abbiamo preparato una [guida passo passo](https://developers.fattureincloud.it/docs/guides/invoice-creation) che le chiederei di provare a seguire. Dovrebbe rispondere alle domande che potrebbero emergere.

Qualora dovesse avere ulteriori domande le chiederei di aprire una nuova discussion, in modo da non andare troppo OT in questa.

A disposizione
Mauro

**Reply by andrealune** [reply]
@valmoz riguardo al secondo punto, come vede sto utilizzando l'ID del payment_account ottenuto dalla chiamata precedente (2048886).
Se ho recuperato il dato sbagliato, la prego di indicarmi dove ottenere il payment_account_id corretto.
Grazie mille

**EDIT: ho capito, stavo confondendo payment_account con payment_method**
